kong/clustering.lua

```diff
@@ -0,0 +1,409 @@
+local _M = {}
+
+
+local semaphore = require("ngx.semaphore")
+local ws_client = require("resty.websocket.client")
+local ws_server = require("resty.websocket.server")
+local ssl = require("ngx.ssl")
+local cjson = require("cjson.safe")
+local declarative = require("kong.db.declarative")
+local utils = require("kong.tools.utils")
+local openssl_x509 = require("openssl.x509")
+local assert = assert
+local setmetatable = setmetatable
+local type = type
+local ipairs = ipairs
+local ngx_log = ngx.log
+local ngx_sleep = ngx.sleep
+local cjson_decode = cjson.decode
+local cjson_encode = cjson.encode
+local kong = kong
+local dc = declarative.new_config(kong.configuration)
+local ngx_exit = ngx.exit
+local exiting = ngx.worker.exiting
+local ngx_time = ngx.time
+local new_tab = require("table.new")
+local ngx_var = ngx.var
+local io_open = io.open
+
+
+local MAX_PAYLOAD = 65536 -- 64KB
+local PING_INTERVAL = 30 -- 30 seconds
+local WS_OPTS = {
+  max_payload_len = MAX_PAYLOAD,
+}
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local ngx_INFO = ngx.INFO
+local WEAK_KEY_MT = { __mode = "k", }
+local CERT_DIGEST
+local CERT, CERT_KEY
+local clients = setmetatable({}, WEAK_KEY_MT)
+local shdict = ngx.shared.kong_clustering -- only when role == "control_plane"
+local prefix = ngx.config.prefix()
+local CONFIG_CACHE = prefix .. "/config.cache.json"
+
+
+local function update_config(config_table, update_cache)
+  assert(type(config_table) == "table")
+
+  local entities, _, _, _, new_hash = dc:parse_table(config_table)
+  if not entities then
+    return nil, "bad config received from control plane"
+  end
+
+  if declarative.get_current_hash() ~= new_hash then
+    -- NOTE: no worker mutex needed as this code can only be
+    -- executed by worker 0
+    local res, err = declarative.load_into_cache_with_events(entities, new_hash)
+    if not res then
+      return nil, err
+    end
+
+    if update_cache then
+      -- local persistence only after load finishes without error
+      local f, err = io_open(CONFIG_CACHE, "w")
+      if not f then
+        ngx_log(ngx_ERR, "unable to open cache file: ", err)
+
+      else
+        local res
+        res, err = f:write(cjson_encode(config_table))
+        if not res then
+          ngx_log(ngx_ERR, "unable to write cache file: ", err)
+        end
+
+        f:close()
+      end
+    end
+
+  else -- get_current_hash() == new_hash
+    ngx_log(ngx_DEBUG, "same config received from control plane,",
+            "no need to reload")
+    return true
+  end
+
+  return true
+end
+
+
+local function send_ping(c)
+  local _, err = c:send_ping(declarative.get_current_hash())
+  if err then
+    ngx_log(ngx_ERR, "unable to ping control plane node: ", err)
+    -- return and let the main thread handle the error
```

Comment:
this comment does not match with the reality. Yes it returns `nil`, but also the good case return `nil` below, and the main thread does not handle the error.

Reply:
The idea here is that the main thread would try to call `recv_frame` which will return the same error, therefore terminating. Since the ping thread already logged the error nothing will be lost. However I could make the ping thread terminate first when this happens. Committing a change.